### ISSUE TITLE
[Feature req]Implement IP-filtering for users

### ISSUE THREAD

**Issue by ict-one-nl** [opensearch-project/security#987]
It would be nice if you can white / black list ip  address blocks (CIDR, multiple include/exclude like the field level security) for specific users to further restrict access to the data. We would use this to limit access for a user (application that gets data from ODFE) to a specific server to further prevent risk of leaking data.

**Comment by cliu123** [comment]
@neographikal Could you elaborate your use cases? Why would you limit access to a specific server? 
In Elasticsearch, data is stored in shards and shards are distributed on different nodes. Query is based on indexes and fields, why does document level security + field level security not help in your use cases?

**Comment by neographikal** [comment]
Document & field level security are fantastic, we use these as well! But implementing ip-filtering provides an additional layer of security to prevent other parties from retrieving information with possibly leaked credentials. Especially when the information is sensitive (think personal data under GDPR) this would be a welcome additional barrier to prevent abuse. It is also less prone to implementation errors should a bug pop up in the filtering functionality.

**Comment by scrawfor99** [comment]
[Triage] Hi @ict-one-nl thank you for filing this issue and re-introducing the conversation, I wanted to follow up on this issue and let you know that for the foreseeable future, we think this service would be best addressed via a network configuration as opposed to something natively in OpenSearch or the Security plugin. There are concerns with the maintenance of a system which could ensure the proper routing and handling of the proxy service.

**Comment by ict-one-nl** [comment]
Hi Stephen, I understand your way of thinking, this was our solution at first as well: solve it in the ingress of Kubernetes. But this is a very rough solution, one size fits all. I'll try to explain it better;

We are using openid for auth
We have 30+ clients talking to opensearch
For each of these clients strict limitations apply because of the GDPR, fls/dls per client
Each client get its own appid/secret from Azure with a role claim for the backend role

If we whitelist AWS or Azure because one client is hosted there, with ip whitelist on Kubernetes, every host in the Azure/AWS blocks could use one of the credentials. Our own proxy unwraps the JWT, get's the appid, looks in the configuration whether *that* client is coming from an ip-range we allow.

Basically if we use a central whitelist, as soon as one client is hosted in the public cloud we would have to whitelist a large portion of the internet. That's quite a large attack surface.

**Comment by scrawfor99** [comment]
Hi @ict-one-nl, I appreciate your detailed explanation for your use case. I can see why you would want this functionality to be native to OpenSearch for your use case. From talking with some of the project maintainers, I think there is concern about the implementation and maintenance process as well as the scope of this feature. It seems like it would be beyond the general scope of the Security plugin to be able to implement this type of proxy support for IP-filtering and it is not clear whether OpenSearch would be able to provide the level of functionality expected when there masked IPs from external sources. I am recalling from memory so this may not be the exact concerns they had but I am going to tag @peternied as someone who may be able to more clearly explain the reasoning.

I appreciate your understanding.

**Comment by peternied** [comment]
@ict-one-nl While building IP based checks are straight forward by using CIDR blocks, the number of edge cases associated with getting the 'correct' IP address is complex.

Our primary concern with building a feature around IP address is that while we could provide this feature for simple network configurations, for more interesting network setups there will be problematic gaps that are hard to troubleshoot.  E.g. If a request goes through a proxy it with an `X-Forwarded-For` header on it, it could either not match to the expected IP or no matter the source IP address the request be considered coming from a trusted configuration.

In environments where such systems are in use, it likely will be easier to grant/denied access using those networking appliances rather than attempting to use the security plugin as a catch all.

**Comment by ict-one-nl** [comment]
Hm I understand what you are trying to say, thanks for the explanation. We are doing a fair amount of reverse proxying within the platform, in my experience you either use the client ip in the request or the x-forwarded-for header. It's up to the reverse proxy to fill this header with the correct information. If you would consider this conditional, take the source ip or use the x-forwarded-for header if set it's clear what can and can not be done.

This would reduce complexity, give flexibility. It would be up to the user/admin to fill the header in a correct manner.

**Comment by peternied** [comment]
> It would be up to the user/admin to fill the header in a correct manner.

I am hesitant to incorporate features that are easy to make mistakes with and then blame the user.  Maybe I'm being over-cautious.  I would love to review a draft PR to see the design elements and use cases.

**Comment by ict-one-nl** [comment]
How would a PR look like? I'm not a native java programmer and the code base is quite large, so implementing this myself would be quite an endeavour :)

**Comment by peternied** [comment]
Everyone starts somewhere, I don't have a good starting place offhand @cwperks would you have a recommendation?

**Comment by wbeckler** [comment]
@ict-one-nl Maybe you could submit pseudocode if that's easier. 

There might be another way to go about this. If the need is to have object-level security combined with identity and IP filtering by role, there are products that exist specifically for that purpose. For example, [this product](https://www.datasparc.com/wp-content/uploads/2019/01/DBHawk-Product-Brief.pdf) claims compatibility with Elasticsearch and with AWS. This company also allows for highly configurable access control and works as a proxy for OpenSearch: https://www.strongdm.com/docs/admin-ui-guide/infrastructure-management/datasources/datasource-amazon-es

**Comment by cwperks** [comment]
@ict-one-nl I see you pointed out the `hosts` feature of `roles_mappings` earlier. I could point you to key files to show you the `hosts` implementation for roles if it can help nudge someone towards a PR for this ;). There are many instances in the security plugin where a feature must be explicitly turned on to be used, so I would be happy to look at a PR for a feature like this.

- Example of a `roles_mapping` with hosts: https://github.com/opensearch-project/security/blob/main/src/test/resources/roles_mapping.yml#L325C16-L337

- `getHosts` is used within `ConfigModelV7` - https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/securityconf/ConfigModelV7.java#L1158

- Roles are only added to the user if the host matches the pattern: https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/securityconf/ConfigModelV7.java#L1238-L1245

I only see WildcardMatchers in the code so to match CIDR a new matcher needs to be introduced. Maybe `spring-security-web`? https://stackoverflow.com/a/41482123

For something like this, I could imagine adding `hosts` onto the User model and then checking the host after authentication here: https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/filter/SecurityRestFilter.java#L197

There's so many different authentication methods that I believe the IP Filtering would need to occur after the user's credentials have been authenticated and the User object is created. 

If the roles mapping solution is suitable and only requires adding CIDR matching then I would recommend that route.

**Comment by cwperks** [comment]
I know that one of the concerns raised around this issue was resolving requests that have been proxied to get the original IP address. There is an XffResolver that is used in this repo that should help solve the problem. https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/http/XFFResolver.java#L64

**Comment by ict-one-nl** [comment]
I don't believe the proxy config should be prohibitive to implement this. Knowledge and time on my part is however, as soon as I have some time to spare I'll give it a go. Could be a few weeks though.
